org/mozilla/jss/nss/SSLFDProxy.c

```diff
@@ -240,3 +240,33 @@ JSSL_SSLFDCertSelectionCallback(void *arg,
     *pRetKey = privkey;
     return SECSuccess;
 }
+
+void
+JSSL_SSLFDHandshakeComplete(PRFileDesc *fd, void *client_data)
+{
+    JNIEnv *env = NULL;
+    jobject sslfd_proxy = (jobject)client_data;
+    jclass sslfdProxyClass;
+    jfieldID handshakeCompleteField;
+
+    if (fd == NULL || client_data == NULL || JSS_javaVM == NULL) {
+        return;
+    }
+
+    if ((*JSS_javaVM)->AttachCurrentThread(JSS_javaVM, (void**)&env, NULL) != JNI_OK || env == NULL) {
+        return;
+    }
+
+    sslfdProxyClass = (*env)->GetObjectClass(env, sslfd_proxy);
+    if (sslfdProxyClass == NULL) {
+        return;
+    }
+
+    handshakeCompleteField = (*env)->GetFieldID(env, sslfdProxyClass,
+                                                "handshakeComplete", "Z");
+    if (handshakeCompleteField == NULL) {
+        return;
+    }
+
+    (*env)->SetBooleanField(env, sslfd_proxy, handshakeCompleteField, JNI_TRUE);
+}
```

Comment:
Just a question here. I see this callback just sets some boolean when the handshake it complete. Would there be any case to have a java method receive this callback as well?

Reply:
For one, there's no callback in the SSLEngine specification, so adding it would be surprising. I don't think there's value in adding it because... 

SSLEngine is non-blocking. At each step of the initial handshake, there's a handshake status. It flips back and forth between Wrap and Unwrap as necessary. When the handshake is finished (i.e., this callback fires), it returns Finished. After that, it returns Not Handshaking. So the caller who would be expecting a handshake complete callback execution could alternatively wait for Finished and execute it then. That's the beauty of non-blocking SSLEngine over a SSLengine: it needs fewer callbacks. :-)

Reply:
And, `SSL_SecurityStatus()` reports `on > 0` even when the handshake itself hasn't completed. This is most commonly when the `FINISHED` message hasn't yet been sent but the server has approved a particular protocol version / cipher suite pair. The `HandshakeCallback` delays firing until the `FINISHED` message is on the wire, making it better than `SecurityStatus.on > 0` (which is what we've replaced in this PR).

Since our call graph looks like:

 - `SSLEngine.wrap` or `SSLEngine.unwrap`
 - `JSSEngine.updateHandshakeState()`
 - `SSL_ForceHandshake()`
 - ... NSS stuff ...
 - This handshake complete handler

We'd prefer to set a boolean and unwind back to `JSSEngine.UpdateHandshakeState` rather than nesting further; the next thing we need to do is in `UpdateHandshakeState` once `SSL_ForceHandshake` returns.